src/Controller/Security/ResetPasswordController.php

```diff
@@ -0,0 +1,170 @@
+        }
+
+        if (!$alreadyUsedHash = $request->query->getString(self::ALREADY_USED_HASH_PARAM)) {
+            throw $this->createNotFoundException('Already used hash not set');
+        }
+
+        if (!hash_equals($this->computeAlreadyUsedHash($user), $alreadyUsedHash)) {
+            $this->addFlash('error', 'This password reset link has already been used.');
+
+            return $this->redirectToRoute('password_reset_request');
+        }
+
+        $form = $this->createForm(ChangePasswordForm::class);
+        $form->handleRequest($request);
+
+        if ($form->isSubmitted() && $form->isValid()) {
+            /** @var string $plainPassword */
+            $plainPassword = $form->get('plainPassword')->getData();
+
+            $user->setPassword($this->passwordHasher->hashPassword($user, $plainPassword));
+            $this->em->flush();
+            $request->getSession()->remove(self::SESSION_ID);
+            $this->addFlash('success', 'Your password has been reset successfully, you are now logged in.');
+
+            return $this->security->login($user, 'form_login');
+        }
+
+        return $this->render('security/reset_password.html.twig', [
+            'form' => $form,
+            'user' => $user,
+        ]);
+    }
+
+    private function addIdToSessionAndRedirect(Request $request, int $id): Response
+    {
+        if (!$this->uriSigner->checkRequest($request)) {
+            $this->addFlash('error', 'The link is invalid or has expired, please try again.');
+
+            return $this->redirectToRoute('password_reset_request');
+        }
+
+        $request->getSession()->set(self::SESSION_ID, $id);
+
+        return $this->redirectToRoute('password_reset', [self::ALREADY_USED_HASH_PARAM => $request->query->get(self::ALREADY_USED_HASH_PARAM)]);
+    }
+
+    private function sendResetLink(string $email): void
+    {
+        if (!$user = $this->users->findOneBy(['email' => $email])) {
+            return;
+        }
+
+        $expires = now()->modify(sprintf('+%d seconds', self::EXPIRES_AFTER));
+        $link = $this->uriSigner->sign(
+            $this->generateUrl(
+                'password_reset',
+                [
+                    'id' => $user->getId(),
+                    self::ALREADY_USED_HASH_PARAM => $this->computeAlreadyUsedHash($user),
+                ],
+                UrlGeneratorInterface::ABSOLUTE_URL
+            ),
+            $expires
+        );
+        $message = (new TemplatedEmail())
+            ->to($user->getEmailAddress())
+            ->subject('Password Reset Request')
+            ->htmlTemplate('email/reset_password.html.twig')
+            ->context([
+                'user' => $user,
+                'link' => $link,
+                'expires' => $expires,
+            ])
+        ;
+        $message->getHeaders()->add(new TagHeader('forgot-password'));
+        $message->getHeaders()->add(new MetadataHeader('link', $link));
```

Comment:
should we explain these? Is the tag header useful if the user isn't aware of it?